<a href="https://colab.research.google.com/github/edaska/Stochastic_Processes_-_Optimization_in_Machine_Learning/blob/main/lab2/Stochastic_Processes_%26_Optimization_in_Machine_Learning_(Lab_2_PCA).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1><b><i>Principal Component Analysis</i> (<i>PCA</i>) </b></h1> <p>In this exercise you will examine the <b><i>principal component analysis</i></b> algorithm (<b><i>Principal Component Analysis</i></b>, <b><i>PCA</i></b>), implemented according to the <b><i>covariance method</i></b> (<b><i>covariance method</i></b>). To understand the usefulness of the method, you will train and evaluate the accuracy of a <b><i>logistic regression</i></b> model on a dataset before and after applying the <b><i>PCA</i></b> algorithm. More information about this method can be found <a href="https://ourarchive.otago.ac.nz/handle/10523/7534">here</a>.</p> <p>The exercise includes <b><i>two</i></b> <i>Python</i> programs: (a) the first takes a dataset in .<i>csv</i> format, applies the <b><i>PCA</i></b> algorithm, and creates the file <b><i>foo.csv</i></b> with the transformed dataset, as it results from the principal components that the user chose to keep, and (b) the second takes a file in .<i>csv</i> format, splits the dataset into a <i>training</i> and a <i>test set</i>, trains a <b><i>logistic classification</i></b> model using the <i>training set</i>, and computes the number of model errors on the <i>test set</i>.</p> <p>The dataset you will use is provided in two forms: (a) <b><i><a href="https://github.com/netmode/Stochastic-Processes-and-Optimization-in-Machine-Learning-Lab/blob/main/lab2/demo3a.csv">demo3a.csv</a></i></b> and (b) <b><i><a href="https://github.com/netmode/Stochastic-Processes-and-Optimization-in-Machine-Learning-Lab/blob/main/lab2/demo3b.csv">demo3b.csv</a></i></b>, which does not include the first column of <b><i><a href="https://github.com/netmode/Stochastic-Processes-and-Optimization-in-Machine-Learning-Lab/blob/main/lab2/demo3a.csv">demo3a.csv</a></i></b>, i.e., the <i>labels</i> that correspond to each input. These datasets are a simplified form of the dataset that can be found <a href="https://archive.ics.uci.edu/ml/datasets/wine">here</a>.</p> <h3><b><i>Principal Component Analysis</i></b></h3> <p>First, you will load the libraries required for the program that will analyze the dataset <b><i><a href="https://github.com/netmode/Stochastic-Processes-and-Optimization-in-Machine-Learning-Lab/blob/main/lab2/demo3b.csv">demo3b.csv</a></i></b> into its principal components.</p>




In [1]:

#https://machinelearningmastery.com/calculate-principal-component-analysis-scratch-python/
from numpy import genfromtxt
from numpy import mean
from numpy import cov
from numpy.linalg import eig
import numpy as np

<p>Next, you will load the dataset <b><i></i></b></p>

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
data = genfromtxt('/content/drive/MyDrive/NTUA FINANCIAL ENGINEERING MASTER/2nd Semester/Stochastic Processes & Optimization in Machine Learning/Lab2/demo3b.csv', delimiter=',')

In [5]:
data

array([[1.423e+01, 1.710e+00, 2.430e+00, ..., 1.040e+00, 3.920e+00,
        1.065e+03],
       [1.320e+01, 1.780e+00, 2.140e+00, ..., 1.050e+00, 3.400e+00,
        1.050e+03],
       [1.316e+01, 2.360e+00, 2.670e+00, ..., 1.030e+00, 3.170e+00,
        1.185e+03],
       ...,
       [1.179e+01, 2.130e+00, 2.780e+00, ..., 9.700e-01, 2.440e+00,
        4.660e+02],
       [1.237e+01, 1.630e+00, 2.300e+00, ..., 8.900e-01, 2.780e+00,
        3.420e+02],
       [1.204e+01, 4.300e+00, 2.380e+00, ..., 7.900e-01, 2.570e+00,
        5.800e+02]])

<p>Then, you will compute the mean of each dataset column (feature) and normalize each feature using it</p>

In [6]:
M = mean(data.T, axis=1)
data_normal = data - M

<p>Next, you will compute the <b><i>covariance matrix</i></b> for the dataset</p>

In [7]:
covariance = cov(data_normal.T)
print("The covariance matrix of the normalized data is the following: ")
print(covariance)

The covariance matrix of the normalized data is the following: 
[[ 7.89911157e-01  1.64350328e-02  5.03965534e-02 -1.39557561e+00
   4.51870543e+00  2.35123047e-01  3.56524806e-01 -2.92133512e-02
   9.15320632e-02  1.07205560e+00  3.92238044e-03  1.22678110e-01
   2.36909466e+02]
 [ 1.64350328e-02  7.74344335e-01  3.18039952e-02  4.24165474e-01
  -2.21434109e-01  1.44160704e-02  3.75581395e-02  5.34385808e-03
   6.47891175e-02 -1.54797734e-01 -6.06599761e-02  6.96007454e-02
  -3.33629219e+01]
 [ 5.03965534e-02  3.18039952e-02  8.83052117e-02  3.73006798e-01
   1.40136434e+00  4.12452594e-02  8.26573643e-02  5.81539654e-03
   1.24709839e-02  1.22397007e-01  2.06346094e-03  3.02871079e-02
   3.18556160e+01]
 [-1.39557561e+00  4.24165474e-01  3.73006798e-01  1.15631181e+01
  -6.46193798e+00 -4.24464222e-01 -4.52046512e-01  1.24282469e-01
  -1.77910614e-01 -2.39565069e+00 -2.05387955e-02 -1.02069171e-03
  -5.11872284e+02]
 [ 4.51870543e+00 -2.21434109e-01  1.40136434e+00 -6.46193798e+00
  

<p>The next step is to compute the <b><i>eigenvalues</i></b> and <b><i>eigenvectors</i></b> of the dataset.</p>

In [8]:
values, vectors = eig(covariance)
print("The eigenvalues of the normalized data are the following: ")
print(values)

The eigenvalues of the normalized data are the following: 
[1.24313073e+05 1.93456026e+02 9.50168321e+00 1.32500169e+00
 8.10595669e-01 4.51081713e-01 2.69157831e-01 1.60993380e-01
 1.11100343e-01 6.13399000e-03 2.19782528e-02 3.20312852e-02
 5.87421686e-02]


In [14]:
#New lines of code to sort the eigenvalues in descending order, as requested in the question
sorted_indices = np.argsort(values)[::-1]
values = values[sorted_indices]

print("The sorted eigenvalues of the normalized data are the following: ")
print(values)

The eigenvalues of the normalized data are the following: 
[1.24313073e+05 1.93456026e+02 9.50168321e+00 1.32500169e+00
 8.10595669e-01 4.51081713e-01 2.69157831e-01 1.60993380e-01
 1.11100343e-01 5.87421686e-02 3.20312852e-02 2.19782528e-02
 6.13399000e-03]


<p>Next, you will select the most important <b><i>eigenvalues</i></b> and adjust the corresponding <b><i>eigenvectors</i></b> of the dataset accordingly.</p>

In [9]:
new_values = values[0:3]
print("The most important eigenvalues are the following: ")
print(new_values)
new_vectors = vectors[0:3]
print("The most important eigenvectors are the following: ")
print(new_vectors)

The most important eigenvalues are the following: 
[1.24313073e+05 1.93456026e+02 9.50168321e+00]
The most important eigenvectors are the following: 
[[-1.90615566e-03 -4.75319302e-04 -4.70065618e-02  2.25255930e-01
   9.82089183e-02 -2.20758056e-01 -8.52483156e-01 -3.65140031e-01
  -1.65925760e-01 -5.62139324e-03 -2.07811823e-02  2.48378617e-02
  -8.18675258e-04]
 [ 2.68379014e-04 -2.11118304e-03  3.21935310e-02  1.25115825e-01
   9.16405147e-01 -3.07909300e-01  1.86306127e-01  7.26209733e-02
  -1.81047372e-02 -1.16925805e-02  8.44468784e-02 -2.41209813e-02
   2.31770029e-02]
 [-2.56407459e-04 -4.21644285e-03  5.19948847e-02  4.67286333e-02
   2.01302236e-02 -3.45064791e-03 -1.57617050e-02  1.52544206e-01
  -7.18891029e-02 -1.38330530e-01 -6.68720003e-02  6.75337422e-01
  -6.97354965e-01]]


<p>Now, you will apply the new <b><i>eigenvectors</i></b> to the old dataset in order to obtain the new, reduced-size dataset.</p>

In [10]:
new_data = new_vectors.dot(data_normal.T)

In [16]:
new_data

array([[ 1.11059332e+00, -2.04662553e+00, -1.14053934e+00,
        -9.23027719e-01,  2.03011307e+00, -1.15565147e+00,
        -1.77979657e+00,  1.44429215e+00, -2.07754903e+00,
        -1.66158456e+00, -1.13894942e+00, -1.16244734e+00,
        -2.38966598e+00, -4.21662666e+00, -3.30443663e+00,
        -7.20223346e-03,  1.15570174e+00,  5.48901205e-01,
        -1.90211655e+00,  2.97792269e-01,  1.24511780e+00,
         2.63072433e-01, -8.79338438e-01, -6.91895384e-01,
        -2.14239318e-01,  3.43602836e+00, -2.03110979e+00,
        -1.00871394e+00,  1.96208735e-01, -1.09721070e+00,
        -3.38536473e-01, -6.06947370e-01, -3.40746063e-01,
         2.61870549e+00,  8.26058124e-01, -1.55479436e-01,
         2.95322652e-02, -5.10116030e-01, -1.12888398e+00,
         9.46478400e-01,  1.73780173e-01, -1.23587877e+00,
        -2.00144907e+00, -1.51223523e-02, -3.30453965e-01,
         7.63642835e-01, -1.42928844e+00, -1.58366365e+00,
        -4.51982805e-01, -9.48461832e-01, -3.35614037e+0

<p>Save the new dataset to a <i>csv</i> file.</p>

In [11]:
np.savetxt("foo.csv", new_data.T, delimiter=",")

In [12]:
with open("foo.csv", 'r') as f:
    print("\nThe file contains:")
    print(f.read())


The file contains:
1.110593323544648348e+00,3.077698859592911163e+01,-1.909089418861344711e+02
-2.046625533117178986e+00,5.161372563443116235e+00,-1.813155934965849951e+02
-1.140539344223838070e+00,1.015546123023898062e+01,-2.755132612766655598e+02
-9.230277193194514895e-01,2.744447017804328581e+01,-4.811471695644075339e+02
2.030113068849923152e+00,1.555547661974124907e+01,3.887919210094035805e+01
-1.155651466668587979e+00,2.584513102556409692e+01,-4.605619040820255918e+02
-1.779796574261774733e+00,7.468392560543437142e+00,-3.486236956780053902e+02
1.444292151066926877e+00,3.086123006634682753e+01,-3.513809975554720495e+02
-2.077549031831548820e+00,2.638813542109645915e+00,-1.782932409597241019e+02
-1.661584560387775689e+00,3.736125141721999388e+00,-1.779796260499848017e+02
-1.138949421467751133e+00,2.125575446969192228e+01,-5.021272417449745262e+02
-1.162447344883549061e+00,6.706083403045473013e+00,-3.420353970474981793e+02
-2.389665982376205200e+00,1.962426928835504292e+00,-3.701309

<h4><b><i>Questions</i></b></h4> <ul> <li>Check the program above and briefly describe the steps followed by the <b><i>PCA</i></b> algorithm, implemented using the <b><i>covariance</i></b> method. Include the mathematical operations as well.</li> <li>Apply the <b><i>PCA</i></b> algorithm to the data in the file <i><a href="https://github.com/netmode/Stochastic-Processes-and-Optimization-in-Machine-Learning-Lab/blob/main/lab2/demo3b.csv">demo3b.csv</a></i>. Then, record the dataset’s <i>covariance matrix</i> and the eigenvalues of this matrix. What do the positive values and what do the negative values of the <i>covariance matrix</i> indicate? Sort the <i>eigenvalues</i> in descending order. What do you observe about the first three compared to the rest? How many <i>principal components</i> does the algorithm choose to keep?</li> </ul>

The PCA algorithm using the covariance method begins by calculating the mean of each feature ($\bar{x}_j = \frac{1}{n} \sum x_{i,j}$) and subtracting these means from the original dataset $X$ to create a zero-centered matrix ($X_{norm} = X - \bar{x}$). Next, it computes the covariance matrix of the normalized data ($cov(X,Y) = \frac{1}{n-1} \sum (X_{i}- \bar{x})(Y_{i}- \bar{y}$) (we will use $cov$ as $C$) to measure how the features vary together. The algorithm then performs eigendecomposition on this matrix to determine its eigenvalues $\lambda$ and eigenvectors $\mathbf{v}$ ($C\mathbf{v} = \lambda\mathbf{v}$). Instead of sorting, the code simply takes the first three eigenvalues and their corresponding eigenvectors (via the slice [0:3]) to form a projection matrix $W$. Finally, the mean-centered dataset is multiplied by this projection matrix ($X_{new}^T = W X_{norm}^T$) to project and transform the data into a new, reduced-dimensional feature space.

If the value is positive, then that indicates that both dimensions increase together. If the value is negative, then as one dimension increases, the other decreases.
Because this line "values, vectors = eig(covariance)" does not do sort, we implemented a new line for sorting. It seemed that the sorted version has the same 3 first eigenvalues as the code before.

1.24313073e+05 = 124,313.073

1.93456026e+02 = 193.456026

9.50168321e+00 = 9.50168321

We can observe that these first three eigenvalues are significantally larger than the others.
Based on the provided code, the algorithm chooses to keep 3 principal components.

<h3><b><i>Logistic Classification</i></b></h3> <p>First, you will load the necessary libraries.</p>

In [ ]:
!pip install pandas
!pip install numpy
!pip install matplotlib
!pip install sklearn
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

Next, you will load the datasets that will be used to train the <b><i>logistic classification</i></b> model. In the first case, you will train the model using the file <i><a href="https://github.com/netmode/Stochastic-Processes-and-Optimization-in-Machine-Learning-Lab/blob/main/lab2/demo3a.csv">demo3a.csv</a></i>. In the second case, you will load the file <i>foo.csv</i> that you obtained as output from the previous code section of the exercise (<i>Principal Component Analysis</i>), adding in the first column of the file the labels that exist in the first column of the file <a href="https://github.com/netmode/Stochastic-Processes-and-Optimization-in-Machine-Learning-Lab/blob/main/lab2/demo3a.csv"><i>demo3a.csv</i></a>.

In [ ]:
df = pd.read_csv("https://github.com/netmode/Stochastic-Processes-and-Optimization-in-Machine-Learning-Lab/blob/main/lab2/demo3a.csv")
#df = pd.read_csv("foo.csv")

<p>Then, you will split the dataset into a <i>training</i> and a <i>test set</i>.</p>

In [ ]:
# Separate the input features from the target variable
x = df.iloc[:,1:13].values
y = df.iloc[:,0].values
# Split the dataset into train and test set
Xtrain,Xtest,Ytrain,Ytest = train_test_split(x,y, test_size = 0.2)

<p>Then, you will train the <b><i>Logistic Classification</i></b> model.</p>

In [ ]:
classifier = LogisticRegression(max_iter = 1000)
classifier.fit(Xtrain,Ytrain)

<p>Finally, you will obtain the predictions of the model you trained on the test set and compute the total number of errors.</p>

In [ ]:
# Get the predictions on the test set
prediction = classifier.predict(Xtest)

# Calculate the total number of errors on the test set
errors = 0
for index in range(0,len(prediction) - 1):
	if prediction[index] != Ytest[index]:
		errors += 1

print("Total errors on the test dataset")
print(errors)

<h4><b><i>Question</i></b></h4> <p>Run the above code segments using as input the files (a) <i><a href="https://github.com/netmode/Stochastic-Processes-and-Optimization-in-Machine-Learning-Lab/blob/main/lab2/demo3a.csv">demo3a.csv</a></i> and (b) <i>foo.csv</i>. What do you observe about the model’s accuracy in the two cases? Also try the case where we keep (a) 1 and (b) 2 <i>principal components</i>. What do you observe?</p>

<h3><b><i>Additional Questions</i></b></h3> <ul> <li>What is the usefulness of the <b><i>PCA</i></b> algorithm with respect to the ability to <i>visualize</i> the dataset’s data?</li> <li>What is the usefulness of the <b><i>PCA</i></b> algorithm with respect to the training speed of the logistic classification model? Base your answer on the following two code segments.</li> </ul>